ApiData.ts

```diff
@@ -0,0 +1,209 @@
+					this.firstVal = from;
+					// Calculate lastVal based on pageSize
+					this.lastVal = from + this.pageSize - 1;
+					this.displayRecords();
+					this.updatePageInfo();
+				})
+				.catch(() => {
+					throw ('Failed to search value');
+				});
+		} else {
+			alert('Please enter values in the range (0-999999)');
+			return Promise.resolve();
+		}
+	}
+
+	/**  use Ajax for data fetching*/
+	private async fetchData(url: string): Promise<number | string> {
+		$('#overlay').show();
+		const response = await $.ajax({
+		  url,
+		  method: 'GET',
+		});
+		$('#overlay').hide();
+		return response;
+	}
+
+	/**change grid height according to screen size*/
+	private adjustGridHeight(): void {
+		const gridElement = document.getElementById('grid');
+		const pageCntrl = $('.grid-controls').innerHeight();
+		const screenHeight = $(window).innerHeight();
+		if (gridElement && pageCntrl !== undefined && screenHeight !== undefined) {
+			this.maxGridHeight = screenHeight - pageCntrl;
+			gridElement.style.height = `${this.maxGridHeight}px`;
+		};
+	}
+
+	/** Update the page information and records display based on the current state of the grid.*/
+	private updatePageInfo(): void {
+		const totalPages = Math.ceil(this.totalItems / this.pageSize);
+		const pageInfo = `Page ${this.currentPage} of ${totalPages}`;
+		const maxRange = this.totalItems - 1;
+		const from = this.firstVal;
+		let to = Math.min(from + this.pageSize, maxRange);
+		$('#pageInfo').text(`${pageInfo}`);
+		$('.records').text(`Showing records ${from} to ${to}`);
+	}
+
+	
+	private setupControls(): void {
+		$('#prevBtn').on('click', () => this.handlePageChange(-1));
+		$('#nextBtn').on('click', () => this.handlePageChange(1));
+		$(window).on('resize', debounce(this.handleResize, 100));
+	}
+
+	private handlePageChange(delta: number): void {
+
+		if (delta > 0 && this.firstVal + delta * this.pageSize > this.totalItems - 1) {
+			this.firstVal = 0;
+		} else if (delta < 0 && this.firstVal + delta * this.pageSize < 0) {
+			this.firstVal = Math.max(0, this.totalItems - this.pageSize);
+		} else {
+			this.firstVal = Math.max(0, Math.min(this.firstVal + delta * this.pageSize, this.totalItems - 1));
+		}
+
+		this.lastVal = this.firstVal + this.pageSize ;
+		this.currentPage = Math.floor(this.firstVal / this.pageSize) + 1;
+		this.fetchRecords();
+		this.updatePageInfo();
+	}
+
+	private handleResize = (): void => {
+		const newGridSize = Math.floor((Math.floor(<number>($(window).innerHeight())) * GRID_RATIO) / ROW_HEIGHT);
+
+		// Check if the new grid size is non-negative
+		if (newGridSize >= 0) {
+			// Adjust firstVal for the last page
+			if (this.firstVal + newGridSize > this.totalItems) {
+				this.firstVal = Math.max(this.totalItems - newGridSize);
+			}
+
+			this.pageSize = newGridSize;
+			this.lastVal = this.firstVal + newGridSize - 1;
+
+			// Fetch records, update page info, and adjust grid height
+			this.fetchRecords();
+			this.updatePageInfo();
```

Comment:
I think it would be better if you only did `updatePageInfo` after `fetchRecords` has finished. Just in case it fails.